File: examples/c-event-kernel/examples/benchmark.c
/**
 * C Event Kernel Benchmark
 * For comparison with Rust implementation
 */

#define _POSIX_C_SOURCE 199309L

#include "../src/process.h"
#include "../src/clock.h"
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define BENCH_EVENT 0x20

/* Benchmark: Event posting and delivery */
static volatile unsigned int received_count = 0;

PROCESS(receiver_process, "receiver");

PROCESS_THREAD(receiver_process, ev, data) {
    PROCESS_BEGIN();

    static unsigned int iterations = 10000;

    while(received_count < iterations) {
        PROCESS_WAIT_EVENT();
        if(ev == BENCH_EVENT) {
            received_count++;
        }
    }

    PROCESS_END();
}

void bench_event_posting(void) {
    printf("\n=== Benchmark: Event Posting & Delivery (C) ===\n");

    process_init();
    clock_init();

    unsigned int iterations = 10000;
    received_count = 0;

    process_start(&receiver_process, NULL);

    /* Process INIT event */
    process_run();

    /* Benchmark event posting */
    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC, &start);

    for(unsigned int i = 0; i < iterations; i++) {
        process_post(&receiver_process, BENCH_EVENT, (void*)(uintptr_t)i);
    }

    clock_gettime(CLOCK_MONOTONIC, &end);

    double post_time = (end.tv_sec - start.tv_sec) +
                       (end.tv_nsec - start.tv_nsec) / 1000000000.0;

    /* Benchmark event processing */
    clock_gettime(CLOCK_MONOTONIC, &start);

    while(process_run() > 0) {
        /* Process events */
    }

    clock_gettime(CLOCK_MONOTONIC, &end);

    double process_time = (end.tv_sec - start.tv_sec) +
                          (end.tv_nsec - start.tv_nsec) / 1000000000.0;

    printf("  Events posted:    %u\n", iterations);
    printf("  Events processed: %u\n", received_count);
    printf("  Post time:        %.6f s\n", post_time);
    printf("  Process time:     %.6f s\n", process_time);
    printf("  Post rate:        %.0f events/sec\n", iterations / post_time);
    printf("  Process rate:     %.0f events/sec\n", received_count / process_time);
    printf("  Avg latency:      %.2f µs/event\n",
           (process_time * 1000000) / received_count);
}

/* Benchmark: Process switching */
static volatile unsigned int yield_count1 = 0;
static volatile unsigned int yield_count2 = 0;
static volatile unsigned int yield_count3 = 0;

PROCESS(yield_process1, "yield1");
PROCESS(yield_process2, "yield2");
PROCESS(yield_process3, "yield3");

PROCESS_THREAD(yield_process1, ev, data) {
    PROCESS_BEGIN();

    static unsigned int iterations = 1000;

    for(unsigned int i = 0; i < iterations; i++) {
        yield_count1++;
        PROCESS_YIELD();
    }

    PROCESS_END();
}

PROCESS_THREAD(yield_process2, ev, data) {
    PROCESS_BEGIN();

    static unsigned int iterations = 1000;

    for(unsigned int i = 0; i < iterations; i++) {
        yield_count2++;
        PROCESS_YIELD();
    }

    PROCESS_END();
}

PROCESS_THREAD(yield_process3, ev, data) {
    PROCESS_BEGIN();

    static unsigned int iterations = 1000;

    for(unsigned int i = 0; i < iterations; i++) {
        yield_count3++;
        PROCESS_YIELD();
    }

    PROCESS_END();
}

void bench_process_switching(void) {
    printf("\n=== Benchmark: Process Switching (C) ===\n");

    process_init();
    clock_init();

    yield_count1 = yield_count2 = yield_count3 = 0;

    process_start(&yield_process1, NULL);
    process_start(&yield_process2, NULL);
    process_start(&yield_process3, NULL);

    /* Process INITs */
    while(process_run() > 0 && (yield_count1 < 1 || yield_count2 < 1 || yield_count3 < 1)) {}

    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC, &start);

    while(process_run() > 0) {
        /* Process yields */
    }

    clock_gettime(CLOCK_MONOTONIC, &end);

    double duration = (end.tv_sec - start.tv_sec) +
                      (end.tv_nsec - start.tv_nsec) / 1000000000.0;

    unsigned int total_yields = yield_count1 + yield_count2 + yield_count3;

    printf("  Total yields:     %u\n", total_yields);
    printf("  Duration:         %.6f s\n", duration);
    printf("  Switch rate:      %.0f switches/sec\n", total_yields / duration);
    printf("  Avg switch time:  %.2f µs\n", (duration * 1000000) / total_yields);
}

/* Benchmark: Memory usage */
void bench_memory_usage(void) {
    printf("\n=== Benchmark: Memory Usage (C) ===\n");
    printf("  clock_time_t:     %zu bytes\n", sizeof(clock_time_t));
    printf("  struct pt:        %zu bytes\n", sizeof(struct pt));
    printf("  struct process:   %zu bytes\n", sizeof(struct process));
    printf("  process_event_t:  %zu bytes\n", sizeof(process_event_t));
}

int main(void) {
    printf("=== Contiki-NG C Event Kernel: Benchmark ===\n");

    bench_memory_usage();
    bench_event_posting();
    bench_process_switching();

    printf("\n=== Benchmarks complete ===\n");

    return 0;
}

File: examples/rust-event-kernel/examples/benchmark.rs
//! Benchmark example
//!
//! Measures performance metrics of the event kernel.

use contiki_ng_rs::prelude::*;
use contiki_ng_rs::platform::SystemClock;
use std::time::Instant;

// Custom event for benchmarking
const EVENT_BENCH: u8 = 0x20;

/// Benchmark event posting and delivery
fn bench_event_posting() {
    println!("\n=== Benchmark: Event Posting & Delivery ===");

    ProcessManager::init();
    SystemClock::init();

    let iterations = 10000;
    let mut received_count = 0;

    let _pid = ProcessManager::start("receiver", async move {
        let _init = wait_event().await;

        for _ in 0..iterations {
            wait_event_type(EVENT_BENCH).await;
        }
    })
    .unwrap();

    // Process INIT
    ProcessManager::run();

    // Benchmark event posting
    let start = Instant::now();

    for i in 0..iterations {
        ProcessManager::post(ProcessId(0), Event::new(EVENT_BENCH, i)).unwrap();
    }

    let post_duration = start.elapsed();

    // Benchmark event processing
    let start = Instant::now();

    while ProcessManager::run() > 0 {
        received_count += 1;
    }

    let process_duration = start.elapsed();

    println!("  Events posted:    {}", iterations);
    println!("  Events processed: {}", received_count);
    println!("  Post time:        {:?}", post_duration);
    println!("  Process time:     {:?}", process_duration);
    println!(
        "  Post rate:        {:.0} events/sec",
        iterations as f64 / post_duration.as_secs_f64()
    );
    println!(
        "  Process rate:     {:.0} events/sec",
        received_count as f64 / process_duration.as_secs_f64()
    );
    println!(
        "  Avg latency:      {:.2} µs/event",
        process_duration.as_micros() as f64 / received_count as f64
    );
}

/// Benchmark timer accuracy
fn bench_timer_accuracy() {
    println!("\n=== Benchmark: Timer Accuracy ===");

    ProcessManager::init();
    SystemClock::init();
    ETimerProcess::init::<SystemClock>();

    let test_cases = vec![
        ("10ms", SystemClock::from_millis(10)),
        ("50ms", SystemClock::from_millis(50)),
        ("100ms", SystemClock::from_millis(100)),
        ("500ms", SystemClock::from_millis(500)),
    ];

    for (name, interval) in test_cases {
        let interval_copy = interval;

        let _pid = ProcessManager::start("timer_test", async move {
            let _init = wait_event().await;

            let start = SystemClock::now();
            delay::<SystemClock>(interval_copy).await;
            let elapsed = SystemClock::now().wrapping_sub(start);

            let expected_ms = interval_copy.as_millis(SystemClock::TICKS_PER_SECOND);
            let actual_ms = elapsed.as_millis(SystemClock::TICKS_PER_SECOND);
            let error = (actual_ms as i32 - expected_ms as i32).abs();

            println!(
                "  {} - Expected: {}ms, Actual: {}ms, Error: {}ms",
                name, expected_ms, actual_ms, error
            );
        })
        .unwrap();

        // Run until complete
        loop {
            let pending = ProcessManager::run();
            if pending == 0 && !ETimerProcess::pending() {
                std::thread::sleep(std::time::Duration::from_millis(1));
                if ProcessManager::run() == 0 {
                    break;
                }
            }
            std::thread::sleep(std::time::Duration::from_micros(100));
        }
    }
}

/// Benchmark process switching overhead
fn bench_process_switching() {
    println!("\n=== Benchmark: Process Switching ===");

    ProcessManager::init();
    SystemClock::init();

    let iterations = 1000;

    for i in 0..3 {
        let _pid = ProcessManager::start(&format!("process_{}", i), async move {
            let _init = wait_event().await;

            for _ in 0..iterations {
                yield_now().await;
            }
        })
        .unwrap();
    }

    // Process all INITs
    for _ in 0..3 {
        ProcessManager::run();
    }

    let start = Instant::now();

    while ProcessManager::run() > 0 {}

    let duration = start.elapsed();

    let total_switches = iterations * 3;
    println!("  Total yields:     {}", total_switches);
    println!("  Duration:         {:?}", duration);
    println!(
        "  Switch rate:      {:.0} switches/sec",
        total_switches as f64 / duration.as_secs_f64()
    );
    println!(
        "  Avg switch time:  {:.2} µs",
        duration.as_micros() as f64 / total_switches as f64
    );
}

/// Benchmark memory usage
fn bench_memory_usage() {
    println!("\n=== Benchmark: Memory Usage ===");

    println!("  ClockTime:        {} bytes", std::mem::size_of::<ClockTime>());
    println!("  Timer:            {} bytes", std::mem::size_of::<Timer>());
    println!("  EventTimer:       {} bytes", std::mem::size_of::<EventTimer>());
    println!("  Event:            {} bytes", std::mem::size_of::<Event>());
    println!("  ProcessId:        {} bytes", std::mem::size_of::<ProcessId>());
    println!("  Semaphore:        {} bytes", std::mem::size_of::<Semaphore>());
    println!(
        "  BinarySemaphore:  {} bytes",
        std::mem::size_of::<BinarySemaphore>()
    );
}

fn main() {
    println!("=== Contiki-NG Rust Event Kernel: Benchmark ===");

    bench_memory_usage();
    bench_event_posting();
    bench_timer_accuracy();
    bench_process_switching();

    println!("\n=== Benchmarks complete ===");
}
